cmd-ref: document new gc behavior (#1023)
iterative/dvc.org: content/docs/command-reference/gc.md

# gc

Remove unused files and directories from <abbr>cache</abbr> or
[remote storage](/doc/command-reference/remote).

## Synopsis

```usage
usage: dvc gc [-h] [-q | -v]
              [-w] [-a] [-T] [--all-commits] [-c] [-r <name>]
              [-f] [-j <number>] [-p [<path> [<path> ...]]]
```

## Description

This command deletes (garbage collects) data files or directories that exist in

> DVC cache but are no longer needed. With `--cloud` it also removes data in

Author: > but are no longer needed.

It'd be better to express this in any other way. I don't have a better suggestion, though than the following:
```suggestion
DVC cache but are no longer in use. With `--cloud` it also removes data in
```

Reviewer: Good question, I think needed was there before. I don't have a very strong opinion ... may be even `needed` better because it is more precise here - I just keep stuff that I need. need is more general then use. Still, need to review other place with in use/need.

Reviewer: Changed to always use `need` vs `in-use`. I think `need` and `needed` are actually more correct here.

[remote storage](/doc/command-reference/remote).

To avoid accidentally deleting data, it raises an error and doesn't touch any
files if no scope options are provided. It means it's user's responsibility to
explicitly provide the right set of options to specify what data is still in-use
(so that DVC can figure out what fils can be safely deleted).

> One of the scope options - `--workspace`, `--all-branches`, `--all-tags`, or

Reviewer 2: ```suggestion
One of the scope options, `--workspace`, `--all-branches`, `--all-tags`,
```

`--all-commits`, or any combination of them must be provided. Each of them
corresponds to the current workspace _and_ a set of commits to analyze what
files, directories and what versions are still needed and should be kept (by

> analyzing DVC-files in those commits).

Reviewer 2: ```suggestion
`--all-commits`, or any combination of them must be provided. Each of them
corresponds to keeping the data for the current workspace and for a
different set of commits (determined by reading the DVC-files in them).
```

Reviewer 2: Question: what's the point of this option? To me it sounds like "don't gc anything" except maybe deleted parts of the repo like removed branches, rewritten commits, etc.

Reviewer: DVC cache contains data that is not referenced in any way in Git history. E.g. run do a lot of `dvc run`, or `dvc repro` after modifying certain files/settings. If `--no-commit` is not specified DVC is saving data to cache not matter you do `git commit` to save DVC-files or not.

Reviewer: We probably mention this.

Reviewer: mentioned a possible use case in the recent version - @jorgeorpinel pls take a look

Reviewer 2: Makes sense, thanks.

Reviewer 2: Still this suggestion wasn't applied but I'll address in regular updates.

Reviewer: I put a new paragraph that explains the use case, have you seen it? Or may be I didn't get you specific suggestion ... 🤔

Reviewer 2: I see it, thanks. May reword a little in regular updates but it's great!

> One of the use cases for this option is to safely delete all temporary data DVC cached when dvc run and/or dvc repro were run without committing changes to DVC-files (thus potentially caching data that is not referenced from workspace or Git commits).

Unless the `--cloud` option is used, `dvc gc` does not remove data files from
any remote. This means that any files collected from the local cache can be
restored using `dvc fetch`, as long as they have previously been uploaded with
`dvc push`.

### Removing data in remote storage

If `--cloud` option is provided, command deletes unused data not only in local
DVC cache, but also in remote storage. It means it can be dangerous since in
most cases removing data locally and in remote storage is irreversible.

The default remote is cleaned (see `dvc config core.remote`) unless the
`--remote` option is used.

## Options

- `-w`, `--workspace` - keep cached objects _only_ referenced in the current
  workspace This option is enabled automatically if `--all-tags`,
  `--all-branches`, or `--all-commits` are used.

- `-a`, `--all-branches` - keep cached objects referenced in all Git branches as
  well as in the workspace (implies `-w`). Useful for keeping data for all the

>   latest experiment versions if branches are used to track those.

Reviewer 2: ```suggestion
  well as in the workspace (implies `-w`). Useful if branches are used to
  track
  different experiments.
```

- `-T`, `--all-tags` - the same as `-a` above, but applies to Git tags as well

>   as in the workspace (implies `-w`). Useful if tags are used to track

Reviewer 2: ```suggestion
  as the workspace (implies `-w`). Useful if tags are used to track
```

  "checkpoints" of an experiment or project. Note that both options can be
  combined, for example using the `-aT` flag.

- `--all-commits` - the same as `-a` or `-T` above, but applies to Git commits
  as well as in the workspace (implies `-w`). Useful for keeping data for all

>   experiment versions ever used in the history of the project.

Reviewer 2: ```suggestion
  as well as the workspace (implies `-w`). Useful for keeping all the data
  used in the entire existing commit history of the project.
```
